include/dfu_client.h

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+/**@file dfu_client.h
+ *
+ * @defgroup dfu_client DFU client for downloading firmware
+ * @{
+ * @brief DFU client interface to connect to a server and download
+ * and apply new firmware to a selected target. Currently, only the
+ * HTTP protocol is supported for download.
+ */
+
+#ifndef DFU_CLIENT_H__
+#define DFU_CLIENT_H__
+
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct dfu_client_object;
+
+/** @brief DFU download status. */
+enum dfu_client_status {
+	/** Indicates that the client was either never connected
+	 *  to the server or is now disconnected. */
+	DFU_CLIENT_STATUS_IDLE      = 0x00,
+	/** Indicates that the client is connected to the server
+	 *  and there is no ongoing download. */
+	DFU_CLIENT_STATUS_CONNECTED = 0x01,
+	/** Indicates that the client is connected to the server
+	 *  and download is in progress. */
+	DFU_CLIENT_STATUS_DOWNLOAD_INPROGRESS = 0x02,
+	/** Indicates that the client is connected to the server
+	 *  and download is complete. */
+	DFU_CLIENT_STATUS_DOWNLOAD_COMPLETE = 0x03,
+	/** Indicates that the firmware download is halted by the
+	 *  application. This status indicates that the application
+	 *  identified a failure when handling the
+	 *  @ref DFU_CLIENT_EVT_DOWNLOAD_FRAG event. */
```

Comment:
Shouldn't this be "one of the @ref dfu_client_evt events"?

Reply:
No. The this state can only be reached if on DFU_CLIENT_EVT_DOWNLOAD_FRAG event, the application decide to NACK using a non zero return code.

Reply:
I see I provided only a partial response.
> I did not understand where we specify where the image is downloaded. That might be configured in one of the undocumented fields in dfu_client_object?

The downloaded image is provided in chucks in each fragment event. So the destination of download is left to the application.


> Also, what is the plan for this library, resp. why is it called DFU client? Currently, it looks like its purpose is only to download any kind of object. In this case, the name is quite misleading. If the plan is to extend it, we should maybe mention that in the documentation already?

This is a generic firmware download client. Generic because it can download images for any target, modem/application. And in future using the same APIs, we should be able to support alternate application protocols like CoAP. A URI parser will enable auto detecting the protocol.

However, right now, we cannot implement all features. 

If there is a preference on the name, please suggest.

Also, note that, we do not really care if what we download is a firmware or not. It could be something else. However, the design is fine tuned for large objects and there is no intention of creating a full fledged generic HTTP client. So, the naming is to suggest we are not creating an alternate HTTP client. And this module, may not be about just HTTP at all in future.

Reply:
> 
> 
> No. The this state can only be reached if on DFU_CLIENT_EVT_DOWNLOAD_FRAG event, the application decide to NACK using a non zero return code.

But then I got it wrong in the conceptual documentation:
![image](https://user-images.githubusercontent.com/11227796/54269350-82406e80-457d-11e9-9d27-934d4a876bbd.png)

Please clarify.

Reply:
We should definitely NOT call it DFU client if it's not a DFU client.

Just Download library? Or Download Manager (that might be a bit over-the-top though, since it's meant to be rather basic).

Reply:
Download what? And as I said, downloading all kind of objects, big, small, type etc require more intelligence. And this we do not plan to support with this module. firmware_download is the intention with this module. So, rename to firmware_download is okay. But I do not wish to expand the scope to download anything generic.

Reply:
We can call it firmware download, but then we also need to restrict it to that in the docs.
Above you wrote "Also, note that, we do not really care if what we download is a firmware or not. It could be something else." - and that's also how I understood it from the input. We can restrict it to firmware, but we cannot call it "firmware download" and then say you can use it for whatever.

Reply:
Ruth, there is no way for me to check if what is being donwloaded is firmware or not. What I am saying is I have tuned the implementation to download firmware images (large objects).

And, I am giving up on this now. Tell me what to do to get an approval, I will do it.

Reply:
download_client?

Reply:
Meh, I'll toss in a few words for brainstorming:
- URL
- Transfer
- File

Perhaps `url_download` ?